### PR TITLE
feat: add aztec-srs

### LINKED ISSUE #1686: Secure SRS for VID
Currently we use jellyfish's `gen_srs_for_testing`, even in production. Obviously, we shouldn't do that.

[EDIT: remove `test-srs` feature flag from `Cargo.toml`, too!]

In an ideal world this SRS would be loaded from a config file. Better yet, a default SRS could be hard-coded so as to save work on io/parse.

### PR DESCRIPTION
**Pull request by alxiong**
Closes #1686 

### This PR: 

- add aztec's SRS behind `aztec-srs` feature flag and conditionally import it

### This PR does not: 

- turn this on by default, as the benchmarking effort by @dailinsubjam should still just use test srs, which makes no difference and could be faster than loading production-ready one from a file

### CODE REVIEW

**Comment by dailinsubjam** [comment]
Thanks for tagging! IIUC, I don't have to turn this on when doing benchmarking, but I have to turn `gpu-vid` on for leaders who need to run vid on GPU right?

**Comment by alxiong** [comment]
> but I have to turn gpu-vid on for leaders who need to run vid on GPU right?

yes, also make sure you have Nvida GPU and have cuda toolkit etc installed. 

The SRS introduced here really doesn't affect your benchmarking effort, just use test-srs for your benchmark

**Comment by ggutoski** [comment]
We should remove `test-srs` feature flag from Cargo.toml, too.

**Comment by alxiong** [comment]
> We should remove test-srs feature flag from Cargo.toml, too.

why? I think we need it most of the time (at least in test) or are you suggesting we only add the flag in `[dev-dependencies]`?

**Comment by ggutoski** [comment]
> are you suggesting we only add the flag in `[dev-dependencies]`?

Yes. It should not be used in production.

**Comment by alxiong** [comment]
Let me document my thinking for future reference

> [!NOTE]
On API design on test v.s. production SRS

AFAICT, there are two approaches: 
1. diff functions optionally enabled via feature flag
    - pros:
      - explicit in names `srs_for_test()`, arguable less error-prone for dev
    - cons:
      - harder to manage combinatorial feature flag conditions when there are more options: `#[cfg(and(not(features = "aztec-srs"), not(feature = "aleo-srs")), ...)`
      - you can't run your tests with `--all-features` in CI, instead you end up specifying all features except the mutually exclusive ones, which is an anti-pattern.
2. same function switched via rustflags
   - pros:
     - ensure mutual exclusive choice via `RUSTFLAGS` (avoiding complicated `cfg` expression)
   - cons (?):
     - different choice means different binaries

An important realization is that **both approaches force downstream users/libraries to follow the same pattern**:
- if you use `test_srs()`, then all downstream libraries would create `my_action_with_test_srs()` separate from `my_action()`, and turn them on optionally via propagating feature flags
- if you use `--cfg crypto_srs = "test"`, then all downstream would need to specify that for the compiler to select the proper code to compile

Remember: 
> features should be additive. That is, enabling a feature should not disable functionality, and it should usually be safe to enable any combination of features.
> -- [Rust book](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification)

IMO, we should adopt option 2. but the unfortunate legacy choice of 1 on Jellyfish means that it may require more refactoring if we decide to switch.
A good example I always refers to is the `backend` or `word size` choice in [dalek's curve25519](https://github.com/dalek-cryptography/curve25519-dalek/tree/main/curve25519-dalek#backends) crate.

cc @mrain @ggutoski @philippecamacho @akonring @chancharles92

**Comment by alxiong** [comment]
Gus, your suggestion in https://github.com/EspressoSystems/HotShot/pull/2854#issuecomment-2030064207 will be cleaner if we switch to aforementioned option 2. but it'll be slightly messier in current "feature-flag" approach. 

for now, we either:
- default to aztec SRS: remove `aztec-srs`, add `test-srs` here which enable `jf-primitives/test-srs` -- this means HotShot users is going to download 60MB by default
- default to test SRS: keep the status quo of this current commit HEAD of this PR.

which do you prefer? @ggutoski

**Comment by alxiong** [comment]
@ggutoski this is ready for review again. lmk if you want further explanation of the upstream update to `ark-srs`, but if we just care about API usage, it should be straightforwardly seen here in this PR.

File: crates/types/src/vid.rs
Comment (jparr721) [inline crates/types/src/vid.rs:139]: nit: failed*

File: crates/types/src/vid.rs
Comment (jparr721) [inline crates/types/src/vid.rs:138]: Could we perhaps remove the magic numbers here? It's unclear what they mean and where they've come from.

File: flake.nix
Comment (jparr721) [inline flake.nix:204]: Is this file always guaranteed to exist?

File: scripts/download_srs_aztec.sh
Comment (jparr721) [inline scripts/download_srs_aztec.sh:18]: Is it wise to use the `gh` tool when not everyone will have it installed? Also, should we pehaps pin this download into a company-owned github repository instead of a personal one?

**Review by jparr721 — CHANGES_REQUESTED** [review]

File: crates/types/src/vid.rs
Comment (mrain) [reply crates/types/src/vid.rs:138]: That's some upperbound that we need to provide here.
Perhaps we should define a constant for it.

File: flake.nix
Comment (alxiong) [reply flake.nix:204]: no, the following line `./download_srs_aztec.sh` will download it if not exists, and skip if already exist, and this check is done every time we enter the dev shell.

File: crates/types/src/vid.rs
Comment (alxiong) [reply crates/types/src/vid.rs:138]: sounds good, will define a constant and add code comment to explain where this number is coming from.

Comment (alxiong) [reply crates/types/src/vid.rs]: done in [47870cb](https://github.com/EspressoSystems/HotShot/pull/2854/commits/47870cb78cbbc2fb81f615f134b1140201ffbc29)

File: crates/types/src/vid.rs
Comment (alxiong) [reply crates/types/src/vid.rs]: fixed in [47870cb](https://github.com/EspressoSystems/HotShot/pull/2854/commits/47870cb78cbbc2fb81f615f134b1140201ffbc29)

File: scripts/download_srs_aztec.sh
Comment (alxiong) [reply scripts/download_srs_aztec.sh]: switch to `wget` and removed `gh` in the `buildInputs` in [47870cb](https://github.com/EspressoSystems/HotShot/pull/2854/commits/47870cb78cbbc2fb81f615f134b1140201ffbc29)

**Comment by ggutoski** [comment]
> @ggutoski this is ready for review again. lmk if you want further explanation of the upstream update to `ark-srs`, but if we just care about API usage, it should be straightforwardly seen here in this PR.

Thanks! Let's talk about it in today's meeting.

**Comment by jparr721** [comment]
> > @ggutoski this is ready for review again. lmk if you want further explanation of the upstream update to `ark-srs`, but if we just care about API usage, it should be straightforwardly seen here in this PR.
> 
> Thanks! Let's talk about it in today's meeting.

Just want to highlight that the build is failing here looks like due to the lazy static poisoning one of the threads. If you guys think it looks good and want to approve I'll be happy to also drop a green check once all the builds are working. Happy to help if needed.

**Comment by alxiong** [comment]
let me fix some of the CI errors, they are caused by the fact that I didn't add `./download_aztec_srs.sh` in their setup

File: .github/workflows/build-and-test.yml
Comment (jparr721) [inline .github/workflows/build-and-test.yml:56]: I wonder if this can be a justfile command or something.

**Review by jparr721 — DISMISSED** [review]
This looks good, but I'd like to wait for @ggutoski or others for confirmation.

File: .github/workflows/build-and-test.yml
Comment (alxiong) [reply .github/workflows/build-and-test.yml:56]: okay! i will do that, I like single responsibility principle.

Comment (jparr721) [reply .github/workflows/build-and-test.yml:56]: I'll re-approve when you add it then!

Comment (alxiong) [reply .github/workflows/build-and-test.yml]: done in [c3fd331](https://github.com/EspressoSystems/HotShot/pull/2854/commits/c3fd33151aecb47f6abaa1df0fa5b3c11052a6fe)

File: crates/types/src/vid.rs
Comment (ggutoski) [inline crates/types/src/vid.rs:134]: Hmm. I did not properly understand until now your dilemma about "same fn" vs "different fn". 😕 I'm a little uneasy with this violation of the additivity of cargo features. My previous understanding is that `test-srs` is purely additive: it would enable a special test srs used to speed up tests, but it would not remove the aztec srs.

What's the reasoning here? Let me guess: we want to save users who run in test mode the trouble of downloading the aztec srs. Is that all? If so then I don't think that's a good enough reason to violate additivity. Test users can suck up the pain and download aztec. 😛

**Review by ggutoski — COMMENTED** [review]
Thanks for all this work! Question below.

File: crates/types/src/vid.rs
Comment (alxiong) [reply crates/types/src/vid.rs:134]: the reason is HotShot seem to want to keep SRS under the same name `static ref KZG_SRS`,
maybe we can have a separate `KZG_SRS_TEST`, which means a separate `fn vid_scheme_test()` etc. and this "extra `_test()` that accepts test srs" pattern will propagate all the way out.

which poison you want to swallow? (I'm fine with both tbh)

File: .github/workflows/build-and-test.yml
Comment (alxiong) [reply .github/workflows/build-and-test.yml]: ah! the env var `AZTEC_SRS_PATH` is needed in the subsequent steps, and setting env var in just command is not possible (afaik), it's only good at reading it or temporarily setting it within the scope of a just command ([see here](https://just.systems/man/en/chapter_37.html))

so I will revert back. lmk if you disagree @jparr721

File: crates/types/src/vid.rs
Comment (ggutoski) [reply crates/types/src/vid.rs:134]: I've consulted others. The mob has spoken. `test-srs` shall be additive, we'll have a separate `KZG_SRS_TEST`.

Comment (alxiong) [reply crates/types/src/vid.rs]: how does this look [39396a1](https://github.com/EspressoSystems/HotShot/pull/2854/commits/39396a1a1778a636c169cd35143bf3c0a36bacb3)?

note that `vid_scheme_for_test()` is not used anywhere currently, but could be in the future if needed.
atm, all test codes are invoking with production SRS, such as `block_contents.rs::precompute_vid_commitment()` which is invoked by `testing/src/block_builder.rs::build_block()`.

^^ it is obviously also fine to run unit test directly on proper SRS, at least for now it's not a memory or speed concern.

File: .github/workflows/build-and-test.yml
Comment (jparr721) [reply .github/workflows/build-and-test.yml:56]: I think that that is fine. Using a `scripts/` folder is a perfectly idiomatic thing as well.

File: crates/types/src/vid.rs
Comment (jparr721) [inline crates/types/src/vid.rs:79]: Since these lines are touched, I want to bring this up: can we remove the newlines and tabs from these logs? It creates non-standard outputs and can cause problems in a structured context.

**Review by jparr721 — CHANGES_REQUESTED** [review]
Quick thing here.

File: crates/types/src/vid.rs
Comment (ggutoski) [reply crates/types/src/vid.rs:79]: no objection so long as Alex is the one to do it. 😆

**Review by ggutoski — DISMISSED** [review]

File: crates/types/src/vid.rs
Comment (alxiong) [reply crates/types/src/vid.rs:79]: what new lines and tabs? you mean line 76?

**Pull request merged**
